### PR TITLE
Use ListFilteredVirtualColumn for left/fact table expression in join condition

### PR DESCRIPTION
**Pull request by rohangarg**
This change allows for usage of `ListFilteredVirtualColumn` if `MV_FILTER` functions are used in a join condition on the left/fact table expression.
<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by cheddar — APPROVED** [review]
LGTM

File: processing/src/main/java/org/apache/druid/query/JoinDataSource.java [hunk 1]
```diff
@@ -221,6 +222,16 @@ public boolean isConcrete()
     return false;
   }
 
+  public Set<String> getVirtualColumnCandidates()
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/JoinDataSource.java:225]: javadoc please

Comment (rohangarg) [reply processing/src/main/java/org/apache/druid/query/JoinDataSource.java:231]: Yes, added it

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 2]
```diff
@@ -161,7 +161,9 @@ public static DruidQuery fromPartialQuery(
   )
   {
     final RelDataType outputRowType = partialQuery.leafRel().getRowType();
-    final VirtualColumnRegistry virtualColumnRegistry = VirtualColumnRegistry.create(sourceRowSignature);
+    final VirtualColumnRegistry virtualColumnRegistry = plannerContext.getVirtualColumnRegistry() == null
+                                                        ? VirtualColumnRegistry.create(sourceRowSignature)
+                                                        : plannerContext.getVirtualColumnRegistry();
```
[the remark this answers is not in the document]
Comment (rohangarg) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java]: As discussed offline, the new changes are now safe wrt the usage of registry between druid queries

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/MultiValueStringOperatorConversions.java [hunk 3]
```diff
@@ -361,6 +363,17 @@ public DruidExpression toDruidExpression(
                 isAllowList()
             )
         );
+
+        if (plannerContext.getVirtualColumnRegistry() != null) {
```
[the remark this answers is not in the document]
Comment (rohangarg) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/MultiValueStringOperatorConversions.java]: Yes, added a comment :+1:

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/PlannerContext.java [hunk 4]
```diff
@@ -392,4 +394,14 @@ public QueryMaker getQueryMaker()
   {
     return Preconditions.checkNotNull(queryMaker, "QueryMaker not available");
   }
+
+  public VirtualColumnRegistry getVirtualColumnRegistry()
```
[the remark this answers is not in the document]
Comment (rohangarg) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/PlannerContext.java]: have changed it to `joinExpressionVirtulaColumnRegistry` which clearer. and added a comment where its getting called

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidJoinQueryRel.java [hunk 5]
```diff
@@ -163,6 +163,9 @@ public DruidQuery toDruidQuery(final boolean finalizeAggregations)
 
     final Pair<String, RowSignature> prefixSignaturePair = computeJoinRowSignature(leftSignature, rightSignature);
 
+    VirtualColumnRegistry virtualColumnRegistry = VirtualColumnRegistry.create(prefixSignaturePair.rhs);
+    getPlannerContext().setVirtualColumnRegistry(virtualColumnRegistry);
```
[the remark this answers is not in the document]
Comment (rohangarg) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidJoinQueryRel.java]: not a problem with the new changes to pass VC registry in the constructor itself

File: processing/src/main/java/org/apache/druid/query/JoinDataSource.java (hunk 1 above)
Comment (rohangarg) [reply processing/src/main/java/org/apache/druid/query/JoinDataSource.java:231]: Yes, added it

**Review by clintropolis — APPROVED** [review]
i think this is ok for now, but should definitely refactor in a follow-up

**Pull request merged**